core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -128,21 +138,83 @@
   }
 
   /**
+   * An interface extending{@link FieldIndex}, represents a unique index.
+   *
+   * @param <T> type of objects in this {@link IndexedSet}
+   */
+  public interface UniqueFieldIndex<T> extends FieldIndex<T> {}
```

Comment:
What happens if someone creates the base class `FieldIndex` instead of one of the more specific ones? It looks like `FieldIndex` is equivalent to `NonUniqueFieldIndex`?

Maybe we could make `FieldIndex` private to prevent users from using that?